src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,717 @@
+
+    source_locationt check_location(loop_head_location);
+    check_location.set_property_class("loop_assigns");
+    check_location.set_comment(
+      "Check assigns clause inclusion for loop " +
+      id2string(check_location.get_function()) + "." +
+      std::to_string(cbmc_loop_id));
+    step_instrs.add(goto_programt::make_assertion(check_var, check_location));
+    step_instrs.add(goto_programt::make_dead(check_var, loop_head_location));
+
+    auto label_instruction =
+      step_instrs.add(goto_programt::make_skip(loop_head_location));
+    goto_instruction->complete_goto(label_instruction);
+  }
+
+  {
+    // Generate havocing code for assigns targets.
+    const auto in_loop_havoc_block =
+      new_tmp_symbol(
+        bool_typet(),
+        loop_head_location,
+        language_mode,
+        symbol_table,
+        std::string(IN_LOOP_HAVOC_BLOCK) + +"__" + std::to_string(cbmc_loop_id))
+        .symbol_expr();
+    step_instrs.add(
+      goto_programt::make_decl(in_loop_havoc_block, loop_head_location));
+    step_instrs.add(
+      goto_programt::make_assignment(in_loop_havoc_block, true_exprt{}));
+    step_instrs.destructive_append(havoc_instrs);
+    step_instrs.add(
+      goto_programt::make_assignment(in_loop_havoc_block, false_exprt{}));
+  }
+
+  {
+    // Assume the loop invariant after havocing the state
+    code_assumet assumption{invariant};
+    assumption.add_source_location() = loop_head_location;
+    converter.goto_convert(assumption, step_instrs, language_mode);
+  }
+
+  {
+    // Generate assignments to store the multidimensional decreases clause's
+    // value just before the loop_head
+    if(!decreases_clauses.empty())
+    {
+      for(size_t i = 0; i < old_decreases_vars.size(); i++)
+      {
+        code_assignt old_decreases_assignment{
+          old_decreases_vars[i], decreases_clauses[i]};
+        old_decreases_assignment.add_source_location() = loop_head_location;
+        converter.goto_convert(
+          old_decreases_assignment, step_instrs, language_mode);
+      }
+    }
+  }
+
+  goto_function.body.destructive_insert(loop_head, step_instrs);
+
+  return step_case_target;
+}
+
+void dfcc_instrument_loopt::add_body_instructions(
+  const std::size_t loop_id,
+  const std::size_t cbmc_loop_id,
+  goto_functionst::goto_functiont &goto_function,
+  symbol_tablet &symbol_table,
+  goto_programt::targett loop_head,
+  goto_programt::targett loop_latch,
+  const exprt &invariant,
+  const exprt::operandst &decreases_clauses,
+  const symbol_exprt &entered_loop,
+  const symbol_exprt &in_base_case,
+  const std::vector<symbol_exprt> &old_decreases_vars,
+  const std::vector<symbol_exprt> &new_decreases_vars,
+  const goto_programt::instructiont::targett &step_case_target,
+  const irep_idt &language_mode)
+{
+  auto loop_head_location(loop_head->source_location());
+  dfcc_remove_loop_tags(loop_head_location);
+  goto_convertt converter(symbol_table, message_handler);
```

Comment:
See various comments below: IMHO the `code_assert assertion{invariant}` is the only case that deserves the use of `goto_convertt`. If that's correct then please move that declaration to the corresponding code block.

Reply:
Correction: the assertion and the use of `decreases_clauses` need it, but none of the earlier ones. So please move it further down, as close as possible to its actual use.